### PR TITLE
Prefs [Task][30726] Client (Firebase Module)

### PR DESCRIPTION
**Pull request by hieuna-441907**
## Related Tickets
- [#Task 30726: Client (Firebase Module)](https://edu-redmine.sun-asterisk.vn/issues/30726)

## WHAT
- Client (Firebase Module)
## HOW
- Client (Firebase Module)
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (HuyTQ)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/appchat/base/BaseViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.sunasterisk.appchat.base
+
+import androidx.lifecycle.ViewModel
+
+abstract class BaseViewModel : ViewModel() {
+    abstract var isLoading: Boolean
```
Comment (datdt-1935) [inline app/src/main/java/com/sunasterisk/appchat/base/BaseViewModel.kt]: dùng `LiveData` đi em

File: app/src/main/java/com/sunasterisk/appchat/db/repository/Repository.kt [hunk 2]
```diff
@@ -1,5 +1,24 @@
...
+import android.net.Uri
+import com.google.firebase.auth.AuthCredential
+import com.google.firebase.auth.AuthResult
+import com.sunasterisk.appchat.db.Result
+import com.sunasterisk.appchat.db.entity.Chat
+import com.sunasterisk.appchat.db.entity.Group
+import com.sunasterisk.appchat.db.entity.User
+import kotlinx.coroutines.flow.Flow
+
 interface Repository {
```
Comment (datdt-1935) [inline app/src/main/java/com/sunasterisk/appchat/db/repository/Repository.kt:12]: tách ra các loại repository khác nhau chứ

File: app/src/main/java/com/sunasterisk/appchat/db/repository/UserRepository.kt [hunk 3]
```diff
@@ -1,17 +1,111 @@
...
+            } else {
+                clientService.getChats(userId).onEach { chatListResult ->
+                    if (chatListResult is Result.Success) {
+                        chatListResult.data.asFlow().buffer()
+                            .onEach { chat ->
+                                val userChatCrossRef =
+                                    UserChatCrossRef(userId = userId, chatId = chat.chatId)
+                                chatDao.insertChat(chat)
+                                userDao.insert(userChatCrossRef)
+                            }.collect()
```
Comment (datdt-1935) [inline app/src/main/java/com/sunasterisk/appchat/db/repository/UserRepository.kt]:
```suggestion
                            }
                            .collect()
```

File: app/src/main/java/com/sunasterisk/appchat/base/BaseViewModel.kt (hunk 1 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/appchat/base/BaseViewModel.kt]: vâng ạ

File: app/src/main/java/com/sunasterisk/appchat/db/repository/Repository.kt (hunk 2 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/appchat/db/repository/Repository.kt:12]: vâng ạ

**Review by thonc-2135 — APPROVED** [review]

**Review by datdt-1935 — APPROVED** [review]

**Pull request merged**